### PR TITLE
feat: Add code blocks transformation to Confluence code macro

### LINKED ISSUE #65: Bug: Code Blocks Rendering as Inline Code in Confluence
# Bug Report: Code Blocks Rendering as Inline Code in Confluence

## Description

When using `@telefonica/markdown-confluence-sync` v2.2.0, all fenced code blocks in Markdown files are rendered as inline code (with backticks) instead of proper Confluence code block macros when synced to Confluence.

## Expected Behavior

Markdown code blocks should be converted to Confluence's `<ac:structured-macro ac:name="code">` storage format with proper language highlighting.

**Example Markdown:**
````markdown
```javascript
function hello() {
    console.log("Hello World");
}
```
````

**Expected Confluence Output:**
```xml
<ac:structured-macro ac:name="code">
  <ac:parameter ac:name="language">javascript</ac:parameter>
  <ac:plain-text-body><![CDATA[function hello() {
    console.log("Hello World");
}]]></ac:plain-text-body>
</ac:structured-macro>
```

## Actual Behavior

Code blocks are rendered as inline code with backticks in the Confluence page body text, losing all formatting and syntax highlighting.

## Root Cause

The library's markdown processing pipeline converts markdown → HTML → Confluence storage format:
- `remark` plugins process markdown and convert to HTML
- `rehype` plugins transform HTML to Confluence storage format

**The Issue:** The pipeline successfully converts markdown code blocks to HTML `<pre><code class="language-x">` tags, but lacks a rehype plugin to convert these HTML elements to Confluence's `<ac:structured-macro>` format.

**File:** `node_modules/@telefonica/markdown-confluence-sync/dist/lib/confluence/transformer/ConfluencePageTransformer.js`

The processing pipeline (lines 52-78) includes plugins for other transformations but is missing a code block converter:
- `rehypeReplaceDetails` - converts HTML details/summary
- `rehypeReplaceStrikethrough` - converts strikethrough
- `rehypeReplaceTaskList` - converts task lists
- **Missing:** Plugin to convert `<pre><code>` to Confluence code macros

## Versions Affected

- `@telefonica/markdown-confluence-sync`: v2.2.0
- Tested on: Node.js v22.18.0
- All programming languages affected (JavaScript, GraphQL, JSON, etc.)

## Reproduction

1. Create a markdown file with a fenced code block:
````markdown
# Test File

```javascript
console.log("test");
```
````

2. Sync to Confluence using the library
3. Observe that code appears as inline code with backticks instead of a proper code block

## Workaround/Fix

We've created a patch that adds the missing rehype plugin:

**New file:** `rehype-confluence-code-blocks.js`
```javascript
// SPDX-FileCopyrightText: 2025 Acoustic, L.P.
// SPDX-License-Identifier: Apache-2.0
import { visit } from 'unist-util-visit';

/**
 * Rehype plugin to convert HTML code blocks to Confluence storage format.
 *
 * Converts:
 *   <pre><code class="language-javascript">code</code></pre>
 *
 * To Confluence storage format:
 *   <ac:structured-macro ac:name="code">
 *     <ac:parameter ac:name="language">javascript</ac:parameter>
 *     <ac:plain-text-body><![CDATA[code]]></ac:plain-text-body>
 *   </ac:structured-macro>
 */
export default function rehypeConfluenceCodeBlocks() {
  return (tree) => {
    visit(tree, 'element', (node, index, parent) => {
      // Look for <pre> elements containing <code>
      if (node.tagName === 'pre' && node.children && node.children.length > 0) {
        const codeNode = node.children[0];

        if (codeNode.type === 'element' && codeNode.tagName === 'code') {
          // Extract language from class (e.g., "language-javascript" -> "javascript")
          let language = '';
          if (codeNode.properties && codeNode.properties.className) {
            const classes = codeNode.properties.className;
            const langClass = classes.find(c => c.startsWith('language-'));
            if (langClass) {
              language = langClass.replace('language-', '');
            }
          }

          // Extract code content
          let code = '';
          const extractText = (node) => {
            if (node.type === 'text') {
              return node.value;
            }
            if (node.children) {
              return node.children.map(extractText).join('');
            }
            return '';
          };
          code = extractText(codeNode);

          // Create Confluence code macro structure
          const confluenceCodeMacro = {
            type: 'element',
            tagName: 'ac:structured-macro',
            properties: {
              'ac:name': 'code',
            },
            children: [],
          };

          // Add language parameter if specified
          if (language) {
            confluenceCodeMacro.children.push({
              type: 'element',
              tagName: 'ac:parameter',
              properties: {
                'ac:name': 'language',
              },
              children: [
                {
                  type: 'text',
                  value: language,
                },
              ],
            });
          }

          // Add code content wrapped in CDATA
          confluenceCodeMacro.children.push({
            type: 'element',
            tagName: 'ac:plain-text-body',
            properties: {},
            children: [
              {
                type: 'raw',
                value: `<![CDATA[${code}]]>`,
              },
            ],
          });

          // Replace the <pre> node with the Confluence macro
          if (parent && typeof index === 'number') {
            parent.children[index] = confluenceCodeMacro;
          }
        }
      }
    });
  };
}
```

**Modification to:** `ConfluencePageTransformer.js`
```javascript
// Add import at top with other rehype plugins (after line 18)
import rehypeConfluenceCodeBlocks from "./support/rehype/rehype-confluence-code-blocks.js";

// Add to processing pipeline (after line 65, after rehypeReplaceTaskList)
.use(rehypeConfluenceCodeBlocks)
```

This fix has been tested and confirmed to work correctly, converting all code blocks to proper Confluence code macros with syntax highlighting.

## Impact

This bug affects all users trying to sync markdown documentation with code examples to Confluence. Technical documentation, API references, and any content with code snippets will display incorrectly.

## Suggested Solution

Add the `rehype-confluence-code-blocks` plugin (or similar functionality) to the library's default rehype processing pipeline in `ConfluencePageTransformer.js`.

## Environment

- OS: macOS (Darwin 25.0.0)
- Node.js: v22.18.0
- Package version: @telefonica/markdown-confluence-sync@2.2.0
- npm: Latest

### PR DESCRIPTION
**Pull request by FrankLedo**
## Summary
Implements proper rendering of markdown code blocks in Confluence by converting them to structured code macros with syntax highlighting support. This addresses issue #65 where code blocks were appearing as inline text instead of formatted code blocks.

## Changes
- Added new `rehype-replace-code-blocks` plugin to transform `<pre><code>` elements to Confluence's code macro format
- Added `confluence.rehype.codeBlocks` configuration option (default: `true`) to enable/disable the feature for compatibility with older Confluence versions
- Comprehensive unit tests with 10 test cases covering various scenarios (empty blocks, multiple languages, special characters, etc.)
- Updated documentation in README.md with configuration details and examples
- Updated CHANGELOG.md under "unreleased" section
- Bumped package version to 2.3.0 (MINOR version per semantic versioning)

## Implementation Details
The plugin:
- Converts markdown fenced code blocks to Confluence's `<ac:structured-macro ac:name="code">` format
- Preserves language syntax highlighting information from code fence language identifiers
- Handles edge cases: empty blocks, blocks without language specification, special characters, multi-line code
- Is enabled by default but can be disabled via configuration for older Confluence instances

## Test Plan
- [x] All existing unit tests pass
- [x] New unit tests added and passing (10 comprehensive tests)
- [x] Linting checks pass
- [x] TypeScript compilation successful
- [x] Feature follows existing plugin patterns in the codebase

Closes #65

### CODE REVIEW

File: components/markdown-confluence-sync/README.md [hunk 1]
```diff
@@ -300,6 +300,8 @@ The namespace for the configuration of this library is `markdown-confluence-sync
 | `confluence.rootPageName` | `string` | Customize Confluence page titles by adding a prefix to all of them for improved organization and clarity | |
 | `confluence.noticeMessage` | `string` | Notice message to add at the beginning of the Confluence pages. | |
 | `confluence.noticeTemplate` | `string` | Template string to use for the notice message. | |
+| `confluence.rehype` | `object` | Rehype plugin options to customize markdown to Confluence HTML conversion. | |
+| `confluence.rehype.codeBlocks` | `boolean` | Enable conversion of code blocks to Confluence code macro format with syntax highlighting. When disabled, code blocks remain as plain HTML pre/code tags. | `true` |
```
Comment (javierbrea) [inline components/markdown-confluence-sync/README.md]: I'd like to propose moving `codeBlocks` to the top level of the configuration instead of nesting it under `confluence`. The reasoning is that this option controls the markdown-to-HTML conversion process, which is the responsibility of this library's core transformation pipeline. The `confluence` namespace should be reserved for options that directly affect the internal `confluence-sync` library—which handles synchronizing already-processed HTML documents with the Confluence instance. This keeps the configuration structure aligned with the architectural boundaries of each component.

**Comment by javierbrea** [comment]
Wow! Fantastic work, thank you so much for this contribution! 🙌 

I really appreciate the thorough implementation and comprehensive test coverage. There are just two small details I'd like to suggest. I've left some detailed comments in the code review regarding the configuration structure and default values.

I'm happy to handle merging this into the release branch and exposing the new configuration option through the general configuration system. I won't ask you to do this part since the code for adding these options can be a bit tricky—and I don't think that's the best use of your time on a first contribution! 😊
